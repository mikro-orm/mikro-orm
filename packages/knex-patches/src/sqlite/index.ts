export * from './SqliteKnexDialect';
export * from './SqliteTableCompiler';
