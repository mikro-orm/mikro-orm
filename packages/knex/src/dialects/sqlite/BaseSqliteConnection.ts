import type { Knex } from 'knex';
import { ensureDir, readFile } from 'fs-extra';
import { dirname } from 'path';
import { AbstractSqlConnection } from '../../AbstractSqlConnection';
import { Utils } from '@mikro-orm/core';

export abstract class BaseSqliteConnection extends AbstractSqlConnection {

  override async connect(): Promise<void> {
    this.createKnex();
    await ensureDir(dirname(this.config.get('dbName')!));
    await this.client.raw('pragma foreign_keys = on');
  }

  getDefaultClientUrl(): string {
    return '';
  }

  override getClientUrl(): string {
    return '';
  }

  override async loadFile(path: string): Promise<void> {
    const conn = await this.client.client.acquireConnection();
    await conn.exec((await readFile(path)).toString());
    await this.client.client.releaseConnection(conn);
  }

  protected override getKnexOptions(type: string): Knex.Config {
    return Utils.mergeConfig({
      client: type,
      connection: {
        filename: this.config.get('dbName'),
      },
      pool: this.config.get('pool'),
      useNullAsDefault: true,
    }, this.config.get('driverOptions'));
  }

  protected transformRawResult<T>(res: any, method: 'all' | 'get' | 'run'): T {
    if (method === 'get') {
      return res[0];
    }

    if (method === 'all') {
      return res;
    }

    if (Array.isArray(res)) {
      return {
        insertId: res[res.length - 1]?.id ?? 0,
        affectedRows: res.length,
        row: res[0],
        rows: res,
      } as T;
    }

    return {
      insertId: res.lastInsertRowid,
      affectedRows: res.changes,
    } as unknown as T;
  }

}
