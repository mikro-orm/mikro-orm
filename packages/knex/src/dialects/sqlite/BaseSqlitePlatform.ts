import { Utils } from '@mikro-orm/core';
import { AbstractSqlPlatform } from '../../AbstractSqlPlatform';

export abstract class BaseSqlitePlatform extends AbstractSqlPlatform {

  override usesDefaultKeyword(): boolean {
    return false;
  }

  override usesReturningStatement(): boolean {
    return true;
  }

  override getCurrentTimestampSQL(length: number): string {
    return super.getCurrentTimestampSQL(0);
  }

  override getDateTimeTypeDeclarationSQL(column: { length: number }): string {
    return 'datetime';
  }

  override getEnumTypeDeclarationSQL(column: { items?: unknown[]; fieldNames: string[]; length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    if (column.items?.every(item => Utils.isString(item))) {
      return 'text';
    }

    /* istanbul ignore next */
    return this.getTinyIntTypeDeclarationSQL(column);
  }

  override getTinyIntTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return this.getIntegerTypeDeclarationSQL(column);
  }

  override getSmallIntTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return this.getIntegerTypeDeclarationSQL(column);
  }

  override getIntegerTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return 'integer';
  }

  override getFloatDeclarationSQL(): string {
    return 'real';
  }

  override getBooleanTypeDeclarationSQL(): string {
    return 'integer';
  }

  override getVarcharTypeDeclarationSQL(column: { length?: number }): string {
    return 'text';
  }

  override convertsJsonAutomatically(): boolean {
    return false;
  }

  override allowsComparingTuples() {
    return false;
  }

  /**
   * This is used to narrow the value of Date properties as they will be stored as timestamps in sqlite.
   * We use this method to convert Dates to timestamps when computing the changeset, so we have the right
   * data type in the payload as well as in original entity data. Without that, we would end up with diffs
   * including all Date properties, as we would be comparing Date object with timestamp.
   */
  override processDateProperty(value: unknown): string | number | Date {
    if (value instanceof Date) {
      return +value;
    }

    return value as number;
  }

  override getIndexName(tableName: string, columns: string[], type: 'index' | 'unique' | 'foreign' | 'primary' | 'sequence'): string {
    if (type === 'primary') {
      return this.getDefaultPrimaryName(tableName, columns);
    }

    return super.getIndexName(tableName, columns, type);
  }

  override getDefaultPrimaryName(tableName: string, columns: string[]): string {
    return 'primary';
  }

  override supportsDownMigrations(): boolean {
    return false;
  }

  override getFullTextWhereClause(): string {
    return `:column: match :query`;
  }

}
