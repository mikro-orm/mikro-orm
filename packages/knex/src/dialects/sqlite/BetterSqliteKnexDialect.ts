import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class BetterSqliteKnexDialect extends MonkeyPatchable.BetterSqlite3Dialect {

  _driver() {
    return require('better-sqlite3');
  }

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

}
