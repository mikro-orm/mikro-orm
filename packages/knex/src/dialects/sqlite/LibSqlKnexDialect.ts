import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class LibSqlKnexDialect extends MonkeyPatchable.BetterSqlite3Dialect {

  get driverName() {
    return 'libsql';
  }

  _driver() {
    return require('libsql');
  }

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

}
