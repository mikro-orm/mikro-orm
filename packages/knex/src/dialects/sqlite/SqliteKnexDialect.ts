import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class SqliteKnexDialect extends MonkeyPatchable.Sqlite3Dialect {

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

  processResponse(obj: any, runner: any) {
    if (obj.method === 'raw' && this.isRunQuery(obj.sql)) {
      return obj.response ?? obj.context;
    }

    return super.processResponse(obj, runner);
  }

  _query(connection: any, obj: any) {
    const callMethod = this.getCallMethod(obj);

    return new Promise((resolve: any, reject: any) => {
      /* istanbul ignore if */
      if (!connection?.[callMethod]) {
        return reject(new Error(`Error calling ${callMethod} on connection.`));
      }

      connection[callMethod](obj.sql, obj.bindings, function (this: any, err: any, response: any) {
        if (err) {
          return reject(err);
        }

        obj.response = response;
        obj.context = this;

        return resolve(obj);
      });
    });
  }

  private getCallMethod(obj: any): string {
    if (obj.method === 'raw') {
      const query = obj.sql.trim().toLowerCase();

      if ((query.startsWith('insert into') || query.startsWith('update ')) && query.includes(' returning ')) {
        return 'all';
      }

      if (this.isRunQuery(query)) {
        return 'run';
      }
    }

    /* istanbul ignore next */
    switch (obj.method) {
      case 'insert':
      case 'update':
        return obj.returning ? 'all' : 'run';
      case 'counter':
      case 'del':
        return 'run';
      default:
        return 'all';
    }
  }

  private isRunQuery(query: string): boolean {
    query = query.trim().toLowerCase();

    if ((query.startsWith('insert into') || query.startsWith('update ')) && query.includes(' returning ')) {
      return false;
    }

    return query.startsWith('insert into') ||
      query.startsWith('update') ||
      query.startsWith('delete') ||
      query.startsWith('truncate');
  }

}
