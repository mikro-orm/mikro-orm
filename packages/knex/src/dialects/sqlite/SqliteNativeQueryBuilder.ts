import { NativeQueryBuilder } from '../../query/NativeQueryBuilder.js';

/** @internal */
export class SqliteNativeQueryBuilder extends NativeQueryBuilder {

  protected override compileTruncate() {
    const sql = `delete from ${this.getTableName()}`;
    this.parts.push(sql);
  }

  protected override addLockClause() {
    return; // not supported
  }

}
