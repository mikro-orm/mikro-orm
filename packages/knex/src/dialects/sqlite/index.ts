export * from './BaseSqliteConnection';
export * from './BaseSqlitePlatform';
export * from './BaseSqliteSchemaHelper';
export * from './SqliteTableCompiler';
export * from './BetterSqliteKnexDialect';
export * from './LibSqlKnexDialect';
export * from './SqliteKnexDialect';
