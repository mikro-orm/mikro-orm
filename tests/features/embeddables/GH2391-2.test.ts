import { Embeddable, Embedded, Entity, MikroORM, OptionalProps, PrimaryKey, Property, UnderscoreNamingStrategy } from '@mikro-orm/sqlite';
import { mockLogger } from '../../helpers';

@Embeddable()
class NestedAudit {

  @Property({ nullable: true, name: 'archivedAt' })
  archived?: Date;

  @Property({ onCreate: () => new Date(), onUpdate: () => new Date() })
  updatedAt!: Date;

  @Property({ onCreate: () => new Date() })
  created!: Date;

}

@Embeddable()
class Audit {

  @Property({ nullable: true, name: 'archivedAt' })
  archived?: Date;

  @Property({ onCreate: () => new Date(), onUpdate: () => new Date() })
  updatedAt!: Date;

  @Property({ onCreate: () => new Date() })
  created!: Date;

  @Embedded(() => NestedAudit)
  nestedAudit1 = new NestedAudit();

}

@Entity()
class MyEntity {

  [OptionalProps]?: 'fooAudit1' | 'barAudit2';

  @PrimaryKey()
  id!: number;

  @Embedded(() => Audit)
  fooAudit1 = new Audit();

  @Embedded(() => Audit, { object: true })
  barAudit2 = new Audit();

}

describe('onCreate and onUpdate in embeddables (GH 2283 and 2391)', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [MyEntity],
      dbName: ':memory:',
      namingStrategy: class extends UnderscoreNamingStrategy {

        propertyToColumnName(propertyName: string, object?: boolean): string {
          if (object) {
            return propertyName;
          }

          return super.propertyToColumnName(propertyName, object);
        }

      },
    });
    await orm.schema.createSchema();
  });

  afterAll(async () => {
    await orm.close(true);
  });

  test('result mapper', async () => {
    expect(orm.em.getComparator().getResultMapper(MyEntity.name).toString()).toMatchSnapshot();
  });

  test(`GH issue 2283, 2391`, async () => {
    const mock = mockLogger(orm, ['query', 'query-params']);
    let line = orm.em.create(MyEntity, {});
    line.fooAudit1.created = new Date(1698010995740);
    line.fooAudit1.updatedAt = new Date(1698010995740);
    line.fooAudit1.nestedAudit1.created = new Date(1698010995740);
    line.fooAudit1.nestedAudit1.updatedAt = new Date(1698010995740);
    line.barAudit2.created = new Date(1698010995740);
    line.barAudit2.updatedAt = new Date(1698010995740);
    line.barAudit2.nestedAudit1.created = new Date(1698010995740);
    line.barAudit2.nestedAudit1.updatedAt = new Date(1698010995740);
    await orm.em.fork().persistAndFlush(line);
    expect(mock).toHaveBeenCalledTimes(3);
    expect(mock.mock.calls[1][0]).toMatch('insert into `my_entity` (`foo_audit1_updated_at`, `foo_audit1_created`, `foo_audit1_nested_audit1_updated_at`, `foo_audit1_nested_audit1_created`, `bar_audit2`) values (1698010995740, 1698010995740, 1698010995740, 1698010995740, \'{"updatedAt":"2023-10-22T21:43:15.740Z","created":"2023-10-22T21:43:15.740Z","nestedAudit1":{"updatedAt":"2023-10-22T21:43:15.740Z","created":"2023-10-22T21:43:15.740Z"}}\') returning `id`');
    mock.mockReset();

    expect(!!line.fooAudit1.created).toBeTruthy();
    expect(!!line.fooAudit1.updatedAt).toBeTruthy();
    expect(!!line.barAudit2.created).toBeTruthy();
    expect(!!line.barAudit2.updatedAt).toBeTruthy();

    line = await orm.em.findOneOrFail(MyEntity, line.id);

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock.mock.calls[0][0]).toMatch('select `m0`.* from `my_entity` as `m0` where `m0`.`id` = 1 limit 1');
    mock.mockReset();

    expect(!!line.fooAudit1.created).toBeTruthy();
    expect(!!line.fooAudit1.updatedAt).toBeTruthy();
    expect(!!line.barAudit2.created).toBeTruthy();
    expect(!!line.barAudit2.updatedAt).toBeTruthy();

    await orm.em.flush();
    expect(mock).not.toHaveBeenCalled();

    jest.useFakeTimers();
    jest.setSystemTime(new Date(1698010995749));
    const tmp1 = line.fooAudit1.archived = new Date(1698010995749);
    await orm.em.flush();
    expect(mock).toHaveBeenCalledTimes(3);
    expect(mock.mock.calls[1][0]).toMatch('update `my_entity` set `foo_audit1_archivedAt` = 1698010995749, `foo_audit1_updated_at` = 1698010995749, `foo_audit1_nested_audit1_updated_at` = 1698010995749, `bar_audit2` = \'{"updatedAt":"2023-10-22T21:43:15.749Z","created":"2023-10-22T21:43:15.740Z","nestedAudit1":{"updatedAt":"2023-10-22T21:43:15.749Z","created":"2023-10-22T21:43:15.740Z"}}\' where `id` = 1');
    mock.mockReset();

    const tmp2 = line.barAudit2.archived = new Date(1698010995750);
    await orm.em.flush();
    expect(mock).toHaveBeenCalledTimes(3);
    expect(mock.mock.calls[1][0]).toMatch('update `my_entity` set `bar_audit2` = \'{"archivedAt":"2023-10-22T21:43:15.750Z","updatedAt":"2023-10-22T21:43:15.749Z","created":"2023-10-22T21:43:15.740Z","nestedAudit1":{"updatedAt":"2023-10-22T21:43:15.749Z","created":"2023-10-22T21:43:15.740Z"}}\' where `id` = 1');
    mock.mockReset();

    const tmp3 = line.barAudit2.nestedAudit1.archived = new Date(1698010995751);
    await orm.em.flush();
    expect(mock).toHaveBeenCalledTimes(3);
    expect(mock.mock.calls[1][0]).toMatch('update `my_entity` set `bar_audit2` = \'{"archivedAt":"2023-10-22T21:43:15.750Z","updatedAt":"2023-10-22T21:43:15.749Z","created":"2023-10-22T21:43:15.740Z","nestedAudit1":{"archivedAt":"2023-10-22T21:43:15.751Z","updatedAt":"2023-10-22T21:43:15.749Z","created":"2023-10-22T21:43:15.740Z"}}\' where `id` = 1');
    mock.mockRestore();

    const line2 = await orm.em.fork().findOneOrFail(MyEntity, line.id);
    expect(line2.fooAudit1.archived).toEqual(tmp1);
    expect(line2.barAudit2.archived).toEqual(tmp2);
    expect(line2.barAudit2.nestedAudit1.archived).toEqual(tmp3);

    jest.useRealTimers();
  });

});
