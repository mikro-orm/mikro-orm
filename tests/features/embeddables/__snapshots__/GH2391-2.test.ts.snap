// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onCreate and onUpdate in embeddables (GH 2283 and 2391) result mapper 1`] = `
"function(result) {
  const ret = {};
  const mapped = {};
  if (typeof result.id !== 'undefined') {
    ret.id = result.id;
    mapped.id = true;
  }
  if (typeof result.foo_audit1 !== 'undefined') {
    ret.fooAudit1 = result.foo_audit1;
    mapped.foo_audit1 = true;
  }
  if (typeof result.foo_audit1_archivedAt !== 'undefined') {
    if (result.foo_audit1_archivedAt == null || result.foo_audit1_archivedAt instanceof Date) {
      ret.foo_audit1_archived = result.foo_audit1_archivedAt;
    } else {
      ret.foo_audit1_archived = new Date(result.foo_audit1_archivedAt);
    }
    mapped.foo_audit1_archivedAt = true;
  }
  if (typeof result.foo_audit1_updated_at !== 'undefined') {
    if (result.foo_audit1_updated_at == null || result.foo_audit1_updated_at instanceof Date) {
      ret.foo_audit1_updatedAt = result.foo_audit1_updated_at;
    } else {
      ret.foo_audit1_updatedAt = new Date(result.foo_audit1_updated_at);
    }
    mapped.foo_audit1_updated_at = true;
  }
  if (typeof result.foo_audit1_created !== 'undefined') {
    if (result.foo_audit1_created == null || result.foo_audit1_created instanceof Date) {
      ret.foo_audit1_created = result.foo_audit1_created;
    } else {
      ret.foo_audit1_created = new Date(result.foo_audit1_created);
    }
    mapped.foo_audit1_created = true;
  }
  if (typeof result.foo_audit1_nested_audit1 !== 'undefined') {
    ret.foo_audit1_nestedAudit1 = result.foo_audit1_nested_audit1;
    mapped.foo_audit1_nested_audit1 = true;
  }
  if (typeof result.foo_audit1_nested_audit1_archivedAt !== 'undefined') {
    if (result.foo_audit1_nested_audit1_archivedAt == null || result.foo_audit1_nested_audit1_archivedAt instanceof Date) {
      ret.foo_audit1_nested_audit1_archived = result.foo_audit1_nested_audit1_archivedAt;
    } else {
      ret.foo_audit1_nested_audit1_archived = new Date(result.foo_audit1_nested_audit1_archivedAt);
    }
    mapped.foo_audit1_nested_audit1_archivedAt = true;
  }
  if (typeof result.foo_audit1_nested_audit1_updated_at !== 'undefined') {
    if (result.foo_audit1_nested_audit1_updated_at == null || result.foo_audit1_nested_audit1_updated_at instanceof Date) {
      ret.foo_audit1_nested_audit1_updatedAt = result.foo_audit1_nested_audit1_updated_at;
    } else {
      ret.foo_audit1_nested_audit1_updatedAt = new Date(result.foo_audit1_nested_audit1_updated_at);
    }
    mapped.foo_audit1_nested_audit1_updated_at = true;
  }
  if (typeof result.foo_audit1_nested_audit1_created !== 'undefined') {
    if (result.foo_audit1_nested_audit1_created == null || result.foo_audit1_nested_audit1_created instanceof Date) {
      ret.foo_audit1_nested_audit1_created = result.foo_audit1_nested_audit1_created;
    } else {
      ret.foo_audit1_nested_audit1_created = new Date(result.foo_audit1_nested_audit1_created);
    }
    mapped.foo_audit1_nested_audit1_created = true;
  }
  if (typeof result.bar_audit2 !== 'undefined') {
    ret.barAudit2 = result.bar_audit2 == null ? result.bar_audit2 : mapEmbeddedResult_0(result.bar_audit2);
    mapped.bar_audit2 = true;
  }
  for (let k in result) { if (result.hasOwnProperty(k) && !mapped[k]) ret[k] = result[k]; }
  return ret;
}"
`;
